File: Data Structures & Algorithms/pacific-atlantic-water-flow/submission-6.py
class Solution:
    def pacificAtlantic(self, grid: List[List[int]]) -> List[List[int]]:
        ROWS, COLS = len(grid), len(grid[0])
        directions = [[1, 0], [-1, 0], [0, 1], [0, -1]]
        pacific = atlantic = False

        def dfs(r, c, prev):
            nonlocal pacific, atlantic
            if r < 0 or c < 0:
                pacific = True
                return

            if r >= ROWS or c >= COLS:
                atlantic = True
                return

            if grid[r][c] > prev:
                return
            
            tmp = grid[r][c]
            grid[r][c] = float('inf')
            for dr, dc in directions:
                dfs(r + dr, c + dc, tmp)
                if pacific and atlantic:
                    break
            grid[r][c] = tmp
        res = []
        for r in range(ROWS):
            for c in range(COLS):
                pacific = False
                atlantic = False
                dfs(r, c, float('inf'))
                if pacific and atlantic:
                    res.append([r, c])
        return res

File: Data Structures & Algorithms/interleaving-string/submission-6.cs
public class Solution {
    public bool IsInterleave(string s1, string s2, string s3) {
        return Dfs(0, 0, 0, s1, s2, s3);
        
    }

    public bool Dfs(int i, int j, int k, string s1, string s2, string s3) {
        if(k == s3.Length) {
            return (i == s1.Length) && (j == s2.Length);
        }
        // if(i > s1.Length && j > s2.Length) {
        //     return false;
        // }

        if(i < s1.Length && s1[i] == s3[k]) {
            if(Dfs(i + 1, j, k + 1, s1, s2, s3)) {
                return true;
            }
        }
        if(j < s2.Length && s2[j] == s3[k]) {
            if(Dfs(i, j + 1, k + 1, s1, s2, s3)) {
                return true;
            }
        } 

        return false;
    }
}
